flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1630,6 +1627,79 @@ public boolean updateState(TaskExecutionState state) {
 		}
 	}
 
+	private boolean updateStateInternal(final TaskExecutionState state, final Execution attempt) {
+		Map<String, Accumulator<?, ?>> accumulators;
+
+		switch (state.getExecutionState()) {
+			case RUNNING:
+				return attempt.switchToRunning();
+
+			case FINISHED:
+				// this deserialization is exception-free
+				accumulators = deserializeAccumulators(state);
+				attempt.markFinished(accumulators, state.getIOMetrics());
+				return true;
+
+			case CANCELED:
+				// this deserialization is exception-free
+				accumulators = deserializeAccumulators(state);
+				attempt.completeCancelling(accumulators, state.getIOMetrics());
+				return true;
+
+			case FAILED:
+				// this deserialization is exception-free
+				accumulators = deserializeAccumulators(state);
+				attempt.markFailed(state.getError(userClassLoader), accumulators, state.getIOMetrics());
+				return true;
+
+			default:
+				// we mark as failed and return false, which triggers the TaskManager
+				// to remove the task
+				attempt.fail(new Exception("TaskManager sent illegal state update: " + state.getExecutionState()));
+				return false;
+		}
+	}
+
+	private void maybeReleasePartitions(final TaskExecutionState state, final Execution attempt) {
+		final ExecutionVertexID finishedExecutionVertex = attempt.getVertex().getID();
+
+		if (state.getExecutionState() == ExecutionState.FINISHED) {
+			final List<IntermediateResultPartitionID> releasablePartitions = partitionReleaseStrategy.vertexFinished(finishedExecutionVertex);
+			releasePartitions(releasablePartitions);
```

Comment:
It might be easy for constructing `ResultPartitionID` here, because we already have `attempt` info in parameters, then we could avoid the complex logic in following `createResultPartitionId`.

Reply:
`attempt` here is the finishing vertex. However, we need the producer of the partition that is incident to the pipelined region.

Reply:
Yes, I missed it. The current attempt is consumer, and we need the correspond producer attempt for the partition.

Reply:
I still felt the following `createResultPartitionId` method seems complex because of involving in many components, I tried to find a way for constructing it easily.

Another options is that `IntermediateResultPartition` could get `ResultPartitionID` directly via internal `IntermediateResultPartitionID` and producer `ExecutionVertex`. Further `SchedulingResultPartition` could also provide the getter method for `ResultPartitionID` via constructing from `IntermediateResultPartition`. I am not sure whether it is worth doing so.

Reply:
Not worth it at the moment because it is not even clear whether it is a good thing to use both `SchedulingTopology` and `FailoverTopology` in the release strategy.

Reply:
We might only need `SchedulingTopology` which was already used in below `createResultPartitionId` method.  It might be like this in `createResultPartitionId`.
`
SchedulingResultPartition schedulingResultPartition = schedulingTopology.getResultPartitionOrThrow(resultPartition);

ResultPartitionID resultPartitonId = schedulingResultPartition.getResultPartitionId();
`
If we add the `ResultPartitionID` info in the constructor of `DefaultSchedulingResultPartition`.